pkg/maintainer/spv/spv.go

```diff
@@ -57,10 +105,348 @@ func (sm *spvMaintainer) startControlLoop(ctx context.Context) {
 }
 
 func (sm *spvMaintainer) maintainSpv(ctx context.Context) error {
-	logger.Infof("Maintaining SPV proof...")
+	for {
+		if err := sm.proveDepositSweepTransactions(); err != nil {
+			return fmt.Errorf(
+				"error while proving deposit sweep transactions: [%v]",
+				err,
+			)
+		}
 
-	// TODO: Implement. For now, just wait.
-	<-ctx.Done()
+		// TODO: Add proving of other type of SPV transactions: redemption
+		// transactions, moving funds transaction, etc.
+
+		select {
+		case <-time.After(sm.config.IdleBackOffTime):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (sm *spvMaintainer) proveDepositSweepTransactions() error {
+	depositSweepTransactions, err := sm.getUnprovenDepositSweepTransactions()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get unproven deposit sweep transactions: [%v]",
+			err,
+		)
+	}
+
+	txProofDifficultyFactor, err := sm.spvChain.TxProofDifficultyFactor()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to get transaction proof difficulty factor: [%v]",
+			err,
+		)
+	}
+
+	// TODO: Handle a situation in which the block headers in the proof span
+	//       multiple Bitcoin difficulty epochs.
+	requiredConfirmations := uint(txProofDifficultyFactor.Uint64())
+
+	for _, transaction := range depositSweepTransactions {
+		accumulatedConfirmations, err := sm.btcChain.GetTransactionConfirmations(
+			transaction.Hash(),
+		)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to get transaction confirmations: [%v]",
+				err,
+			)
+		}
+
+		if accumulatedConfirmations < requiredConfirmations {
+			// Skip the transaction as it has not accumulated enough
+			// confirmations. It will be proven later.
+			continue
+		}
+
+		_, proof, err := bitcoin.AssembleSpvProof(
+			transaction.Hash(),
+			requiredConfirmations,
+			sm.btcChain,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to assemble SPV proof: [%v]", err)
+		}
+
+		// Deserialize the first block header and verify its difficulty.
+		rawBlockHeader := [bitcoin.BlockHeaderByteLength]byte{}
+		copy(
+			rawBlockHeader[:],
+			proof.BitcoinHeaders[:bitcoin.BlockHeaderByteLength],
+		)
+		firstBlockHeader := bitcoin.BlockHeader{}
+		firstBlockHeader.Deserialize(rawBlockHeader)
+		firstBlockHeaderDifficulty := firstBlockHeader.Difficulty()
```

Comment:
What do you think about moving this piece of code to a function exposed by `bitcoin.SpvProof`? We'd use it here as `proof.FirstBlockHeaderDifficulty()`, for example.

Reply:
It makes sense. Done in 7cd1de0b8dbe6cc5ed8db4c8854ad2f46484d89e.